airbyte-integrations/bases/base-python/base_python/client.py

```diff
@@ -0,0 +1,97 @@
+""" MIT License
+
+Copyright (c) 2020 Airbyte
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import inspect
+import json
+import pkgutil
+from abc import ABC
+from datetime import datetime
+from typing import Tuple, Dict
+
+from airbyte_protocol import AirbyteStream, AirbyteRecordMessage
+
+
+class ResourceSchemaLoader:
+    """ JSONSchema loader from package resources
+    """
+    def get_schema(self, name):
+        raw_schema = json.loads(
+            pkgutil.get_data(self.__class__.__module__.split(".")[0],
+                             f"schemas/{name}.json"))
+        return raw_schema
+
+
+class URLSchemaLoader:
+    """ JSONSchema loader from URL
+    """
+    def get_schema(self, name):
+        raise NotImplementedError
+
+
+class BaseClient(ABC):
+    """ Base client for API
+    """
+    schema_loader_class = ResourceSchemaLoader
+
+    def __init__(self, **kwargs):
+        self._schema_loader = self.schema_loader_class()
+        self._stream_methods = self._enumerate_methods()
+
+    def _enumerate_methods(self) -> Dict[str, callable]:
+        """ Detect available streams and return mapping
+        """
+        prefix = 'stream__'
+        mapping = {}
+        methods = inspect.getmembers(self.__class__,
+                                     predicate=inspect.isfunction)
+        for name, method in methods:
+            if name.startswith(prefix):
+                mapping[name[len(prefix):]] = method
+
+        return mapping
+
+    def read_stream(self, stream):
+        """ Yield records from stream
+        """
+        method = self._stream_methods.get(stream.name)
+        if not method:
+            raise ValueError(
+                f"Client does not know how to read stream `{stream.name}`")
+
+        for message in method():
+            now = int(datetime.now().timestamp()) * 1000
+            yield AirbyteRecordMessage(stream=stream.name, data=message,
```

Comment:
I'm not sure I understand what is the usage of this method...
The BaseClient supposed to be used only for our own integrations. In this case, we control how to do state reporting. And as I see it, periodical state report can be done by an intermediate layer that just read records, passes them to actual consumer and periodically emits state (like checkpoints). 
Of course, it could be more complicated if the state message contains all streams that we syncing at this moment.
WDYT?

Reply:
It seems like the layer which outputs the state shouldn't be different from the layer that queries for the records because it is what has knowledge of how the query is performed. If you we delegate state emitting to an intermediate layer, it would leak the implementation detail of how to obtain records from a stream no?

Reply:
@sherifnada I see what you mean here. Can we safely assume that latest record that we read is enough to decide where to continue next sync?

Reply:
Assuming the records are output in the correct order (e.g: ordered ascending on created_at date) that sounds doable